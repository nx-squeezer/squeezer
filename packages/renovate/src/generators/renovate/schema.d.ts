export interface RenovateGeneratorSchema {
  name: string;
  tags?: string;
  directory?: string;
}
