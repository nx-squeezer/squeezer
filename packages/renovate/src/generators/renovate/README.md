# @nx-squeezer/renovate:renovate

Setup command:

```shell
nx g @nx-squeezer/renovate:renovate
```

Shareable config preset for [Renovate](https://www.whitesourcesoftware.com/free-developer-tools/renovate). Angular and Nx dependencies require post upgrade tasks, then Renovate must be configured [self-hosted](https://docs.renovatebot.com/self-hosting/), an easy way to do it would be with [Github Action Renovate](https://github.com/renovatebot/github-action).

You can find an example configuration [here](https://github.com/nx-squeezer/squeezer/blob/main/renovate.config.js) and GitHub workflow [here](https://github.com/nx-squeezer/squeezer/blob/main/.github/workflows/renovate.yml).

For Nx projects it is recommended to use the default preset and Nx monorepo.

This schematic will:

- Configure Renovate to run self-hosted by creating a GitHub Workflow Action.
- Upgrade the CI workflow to run on Renovate branches.
- Add badge to `README.md`/.

Options:

- Force overwriting of configuration with `--force` (default `false`).
- Use [Nx Cloud](https://nx.app) `--useNxCloud` (default `true`).
- Use local presets `--local` (default `false`). This option is not recommended, instead it is better to rely on pointing to the configuration in the GitHub repo `nx-squeezer/squeezer` to always count with the latest version. Use local presets only if you want to customize them.
- Provide the assignee for Renovate PRs with `--assignee=github-user`.

Presets:

<details>
  <summary>Default Preset</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/default.json) includes all presets included in this repo except for [Nx](https://nx.dev/) and [Angular](https://angular.io/) workspaces, they have to be added manually:

```json
"extends": ["github>@nx-squeezer/squeezer"]
```

</details>

<details>
  <summary>Nx Monorepo</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/nxMonorepo.json) groups all dependencies related to [Nx](https://nx.dev/) and [Angular](https://angular.io/), including upgrade schematics:

```json
"extends": ["github>@nx-squeezer/squeezer:nxMonorepo"]
```

> It is incompatible with preset `"github>@nx-squeezer/squeezer:angularWorkspace"`, choose one or another.

</details>

<details>
  <summary>Angular Workspace</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/angularWorkspace.json) groups all dependencies related to [Angular](https://angular.io/), including upgrade schematics:

```json
"extends": ["github>@nx-squeezer/squeezer:angularWorkspace"]
```

> It is incompatible with preset `"github>@nx-squeezer/squeezer:nxMonorepo"`, choose one or another.

</details>

<details>
  <summary>Group All Non Major</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/groupAllNonMajor.json) groups all non-major npm dependencies, including `bump` dependencies:

```json
"extends": ["github>@nx-squeezer/squeezer:groupAllNonMajor"]
```

</details>

<details>
  <summary>Tooling</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/tooling.json) groups all dependencies related to linting and formatting:

```json
"extends": ["github>@nx-squeezer/squeezer:tooling"]
```

Examples:

- `chore(deps): :arrow_up: update dependency`
- `fix(deps): :lock: refresh package-lock.json`
- `fix(deps): :arrow_down: roll back dependency`

</details>

<details>
  <summary>GitHub Workflow Actions</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/githubActions.json) groups all dependencies related to Github actions:

```json
"extends": ["github>@nx-squeezer/squeezer:githubActions"]
```

</details>

<details>
  <summary>Gitmoji Conventional Commits</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/gitmoji.json) adds a :sparkles: gitmoji :sparkles: to the commit message:

```json
"extends": ["github>@nx-squeezer/squeezer:gitmoji"]
```

</details>

<details>
  <summary>NPM</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/npm.json) updates `node` and `npm` only to LTS versions.

```json
"extends": ["github>@nx-squeezer/squeezer:npm"]
```

</details>

<details>
  <summary>Maintenance</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/maintenance.json) configures maintenance of `package-lock.json` file as a separate PR.

```json
"extends": ["github>@nx-squeezer/squeezer:maintenance"]
```

</details>

<details>
  <summary>Widen Range for Library Dependencies</summary>

This [preset](https://github.com/nx-squeezer/squeezer/blob/main/widenRangeLibraryDeps.json) causes Renovate to _not_ update libraries' `package.json`, only the root `package.json` will be updated for minor and patch versions. This has the purpose of keeping libraries with a wider compatibility range, within the last major version. This is the recommended configuration for library authors.

```json
"extends": ["github>@nx-squeezer/squeezer:widenRangeLibraryDeps"]
```

</details>
