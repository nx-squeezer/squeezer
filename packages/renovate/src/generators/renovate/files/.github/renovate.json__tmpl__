{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["config:recommended", ":label(dependencies)", "<% if (local) { %>local><%= gitRepoSlug %><% } else { %>github>nx-squeezer/squeezer<% } %>", "<% if (local) { %>local><%= gitRepoSlug %>:nxMonorepo<% } else { %>github>nx-squeezer/squeezer:nxMonorepo<% } %>", "<% if (local) { %>local><%= gitRepoSlug %>:tooling<% } else { %>github>nx-squeezer/squeezer:tooling<% } %>"<% if (assignee) { %>, ":assignee(<%= assignee %>)"<% } %>]
}
