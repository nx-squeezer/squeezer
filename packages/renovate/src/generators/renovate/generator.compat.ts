import { convertNxGenerator } from '@nx/devkit';

import renovateGenerator from './generator';

/**
 * Angular schematic that adds Renovate to a workspace.
 */
export const renovateSchematic = convertNxGenerator(renovateGenerator);
