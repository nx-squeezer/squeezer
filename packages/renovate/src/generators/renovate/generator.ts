import { formatFiles, Tree } from '@nx/devkit';

import { RenovateGeneratorSchema } from './schema';

export default async function (tree: Tree, options: RenovateGeneratorSchema) {
  await formatFiles(tree);
}
