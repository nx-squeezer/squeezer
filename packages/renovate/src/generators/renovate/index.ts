export * from './generator.compat';
export * as renovateGenerator from './generator';
export * from './renovate';
export * from './schema';
