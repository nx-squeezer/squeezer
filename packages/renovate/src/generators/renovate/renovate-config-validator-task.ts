import { Tree } from '@nx/devkit';

import { exec } from '@nx-squeezer/devkit';

export function renovateConfigValidatorTask(tree: Tree): boolean {
  console.log(`Validating Renovate configuration...`);
  const { output, error } = exec('npx', ['--package', 'renovate', '-c', 'renovate-config-validator'], {
    cwd: tree.root,
  });

  if (error != null) {
    console.error(error);
    return false;
  }

  console.log(output);

  return output.includes(`Config validated successfully`);
}
