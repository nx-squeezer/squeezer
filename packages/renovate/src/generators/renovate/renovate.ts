/**
 * Path of Renovate CI workflow.
 */
export const renovateCiFile = './.github/workflows/renovate.yml';

/**
 * Path of Renovate configuration.
 */
export const renovateConfigFile = '.github/renovate-config.js';

/**
 * Path of Renovate presents.
 */
export const renovateFile = '.github/renovate.json';

/**
 * Path of util script to generate migrations.
 */
export const renovateCreateMigrationsFile = 'nx-create-migrations.sh';

/**
 * List of renovate presets.
 */
export const renovatePresets = [
  'angularWorkspace.json',
  'tooling.json',
  'default.json',
  'githubActions.json',
  'gitmoji.json',
  'groupAllNonMajor.json',
  'maintenance.json',
  'npm.json',
  'nxMonorepo.json',
  'widenRangeLibraryDeps.json',
];

/**
 * Git branch for Renovate PRs.
 */
export const renovateBranch = 'renovate-github/**';
