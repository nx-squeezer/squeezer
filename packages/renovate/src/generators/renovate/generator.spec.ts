import { Tree, readJson } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { parse, stringify } from 'yaml';

import { readmeFile, getGitRepoSlug, securityFile, ciFile } from '@nx-squeezer/devkit';

import { renovateGenerator } from './generator';
import { renovateSchematic } from './generator.compat';
import { renovateCiFile, renovateFile, renovateConfigFile, renovatePresets, renovateBranch } from './renovate';
import { renovateConfigValidatorTask } from './renovate-config-validator-task';

jest.mock('@nx-squeezer/devkit', () => ({
  ...jest.requireActual('@nx-squeezer/devkit'),
  renovateConfigValidatorTask: jest.fn(),
  getGitRepoSlug: jest.fn(),
}));

jest.mock('./renovate-config-validator-task');

describe('@nx-squeezer/renovate renovate generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
    tree.write(ciFile, stringify({ on: { push: { branches: ['main'] } } }));
    tree.write(readmeFile, `# README\n`);
    jest.spyOn(console, 'log').mockImplementation(() => null);
    jest.spyOn(console, 'error').mockImplementation(() => null);
    (getGitRepoSlug as jest.Mock).mockReturnValue('test/test');
  });

  it('should run successfully', async () => {
    await renovateGenerator(tree, { force: true, useNxCloud: true, local: true });
    expect(tree.exists(renovateCiFile)).toBeTruthy();
  });

  it('should provide a schematic', async () => {
    expect(typeof renovateSchematic({ force: true, useNxCloud: true, local: true })).toBe('function');
  });

  it('should run tasks', async () => {
    const tasks = await renovateGenerator(tree, { force: true, useNxCloud: true, local: true });

    expect(tasks).toBeTruthy();

    tasks?.();

    expect(renovateConfigValidatorTask).toHaveBeenCalled();
  });

  it('should skip execution if a Renovate CI workflow already exists', async () => {
    tree.write(renovateCiFile, '');

    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true });

    expect(console.log).toHaveBeenCalledWith(`Renovate workflow already existing at path: ${renovateCiFile}`);
  });

  it('should not skip execution if a Renovate CI workflow already exists but passing force option', async () => {
    tree.write(renovateCiFile, '');

    await renovateGenerator(tree, { force: true, useNxCloud: true, local: true });

    expect(tree.read(renovateCiFile)?.toString()).toBeTruthy();
  });

  it('should generate the Renovate CI workflow', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true });

    const ci = parse(tree.read(renovateCiFile)?.toString() ?? '');

    expect(ci).toMatchSnapshot();
  });

  it('should fail if repo can not be resolved', async () => {
    (getGitRepoSlug as jest.Mock).mockReturnValue(null);

    await expect(
      async () => await renovateGenerator(tree, { force: false, useNxCloud: true, local: true })
    ).rejects.toEqual(new Error(`Could not identify GitHub repo slug.`));
  });

  it('should generate renovate.json with local configuration', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true, assignee: 'user' });

    expect(readJson(tree, renovateFile)).toStrictEqual({
      $schema: 'https://docs.renovatebot.com/renovate-schema.json',
      extends: [
        'config:base',
        ':label(dependencies)',
        'local>test/test',
        'local>test/test:nxMonorepo',
        ':assignee(user)',
      ],
    });
  });

  it('should generate renovate.json with remote configuration', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: false, assignee: 'user' });

    expect(readJson(tree, renovateFile)).toStrictEqual({
      $schema: 'https://docs.renovatebot.com/renovate-schema.json',
      extends: [
        'config:base',
        ':label(dependencies)',
        'github>nx-squeezer/squeezer',
        'github>nx-squeezer/squeezer:nxMonorepo',
        ':assignee(user)',
      ],
    });
  });

  it('should generate renovate-config.js', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true, assignee: 'user' });

    expect(tree.read(renovateConfigFile)?.toString()).toMatchSnapshot();
  });

  it('should generate presets when using local configuration', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true });

    renovatePresets.forEach((preset) => {
      expect(tree.exists(preset)).toBeTruthy();
    });
  });

  it('should not generate presets when using remote configuration', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: false });

    renovatePresets.forEach((preset) => {
      expect(tree.exists(preset)).toBeFalsy();
    });
  });

  it('should fail if can not find GitHub CI file', async () => {
    tree.delete(ciFile);

    await expect(
      async () => await renovateGenerator(tree, { force: false, useNxCloud: true, local: true })
    ).rejects.toEqual(new Error(`Renovate needs a GitHub workflow CI file, none found at: ${ciFile}`));
  });

  it('should add renovate branch to push branches', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: false });

    expect(parse(tree.read(ciFile)?.toString() ?? '').on.push.branches).toContain(renovateBranch);
  });

  it('should add renovate branch to push branches being idempotent', async () => {
    await renovateGenerator(tree, { force: true, useNxCloud: true, local: false });
    await renovateGenerator(tree, { force: true, useNxCloud: true, local: false });

    const ci = parse(tree.read(ciFile)?.toString() ?? '');
    expect(ci.on.push.branches.filter((branch: string) => branch === renovateBranch).length).toEqual(1);
  });

  it('should generate the security file if it did not exist before', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true });

    expect(tree.exists(securityFile)).toBeTruthy();
  });

  it('should skip generating the security file if it existed before', async () => {
    const customSecurity = 'custom';
    tree.write(securityFile, customSecurity);

    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true });

    expect(tree.read(securityFile)?.toString()).toContain(customSecurity);
  });

  it('should add renovate badge', async () => {
    await renovateGenerator(tree, { force: false, useNxCloud: true, local: true });

    expect(tree.read(readmeFile)?.toString()).toContain(
      `![renovate](https://img.shields.io/badge/maintaied%20with-renovate-blue?logo=renovatebot)`
    );
  });
});
