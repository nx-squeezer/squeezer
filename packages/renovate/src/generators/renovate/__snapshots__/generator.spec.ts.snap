// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx-squeezer/renovate renovate generator should generate renovate-config.js 1`] = `
"module.exports = {
  repositories: ['test/test'],
  branchPrefix: 'renovate-github/',
  dryRun: null,
  gitAuthor: 'Renovate Bot GitHub <bot@renovateapp.com>',
  platform: 'github',
  includeForks: false,
  dependencyDashboard: true,
  onboarding: true,
  autodiscover: false,
  allowCustomCrateRegistries: true,
  allowScripts: true,
  exposeAllEnv: true,
  allowPostUpgradeCommandTemplating: true,
  allowedPostUpgradeCommands: [
    '^npm ci --ignore-scripts$',
    '^npm install --ignore-scripts$',
    '^npm run prepare --if-present$',
    '^npm run format --if-present$',
    '^npx --no-install ng update (@[a-z0-9-~][a-z0-9-._~]*\\\\/)?[a-z0-9-~][a-z0-9-._~]* --from=\\\\d+\\\\.\\\\d+\\\\.\\\\d+ --to=\\\\d+\\\\.\\\\d+\\\\.\\\\d+ --migrate-only --allow-dirty --force$',
    '^npx --no-install ng lint --fix$',
    '^npx --no-install nx migrate',
    '^rm -f migrations.json$',
    '^npx --no-install nx run-many --target=lint --all --parallel --fix --skip-nx-cache$',
  ],
};
"
`;

exports[`@nx-squeezer/renovate renovate generator should generate the Renovate CI workflow 1`] = `
{
  "jobs": {
    "renovate": {
      "name": "Renovate",
      "runs-on": "ubuntu-latest",
      "steps": [
        {
          "name": "Checkout repo",
          "uses": "actions/checkout@v3.1.0",
        },
        {
          "env": {
            "LOG_LEVEL": "debug",
            "NX_CLOUD_AUTH_TOKEN": "\${{ secrets.NX_CLOUD_AUTH_TOKEN }}",
          },
          "name": "Self-hosted Renovate",
          "uses": "renovatebot/github-action@v32.217.0",
          "with": {
            "configurationFile": ".github/renovate-config.js",
            "token": "\${{ secrets.RENOVATE_TOKEN }}",
          },
        },
      ],
    },
  },
  "name": "Renovate",
  "on": {
    "schedule": [
      {
        "cron": "0/15 * * * *",
      },
    ],
    "workflow_dispatch": null,
  },
}
`;
